### Comparisons and masking

Just like NumPy allows element-wise arithmetic operations between arrays, for example addition of two arrays, it is also possible to compare two arrays element-wise. For example

In [1]:
import numpy as np

In [2]:
a=np.array([1,3,4])
b=np.array([2,2,7])

In [3]:
c=a < b
print(c)

[ True False  True]


Now we can query whether all comparisons resulted `True`, or whether some comparison resulted `True`:

In [4]:
print(c.all()) # were all True
print(c.any()) # was some comparison True

False
True


We can also count the number of comparisons that were True. This solution relies on the interpretation that True corresponds to 1 and False corresponds to 0:

In [5]:
print(np.sum(c))

2


Because the broadcasting rules apply also to comparison, we can write

In [6]:
print(a > 0)

[ True  True  True]


To try these operations on real data, we download some weather statistics from Helsinki (Kumpula) weather station. More precisely, we will get the daily average temperatures from year 2017. We use the Pandas library, which we will cover later during this course, to load the data.

In [5]:
import pandas as pd
import numpy as np

a=pd.read_csv("https://raw.githubusercontent.com/csmastersUH/data_analysis_with_python_2020/master/kumpula-weather-2017.csv")['Air temperature (degC)'].values

In [2]:
a

array([  0.6,  -3.9,  -6.5, -12.8, -17.8, -17.8,  -3.8,  -0.5,   0.5,
         1.7,  -1.6,  -2.8,   1.1,   0.8,  -2.8,  -4.2,  -3.5,   1.1,
         1.6,  -0.6,  -1.8,   1. ,   0.1,  -2.2,  -3.8,   1.9,   1.6,
         0.8,   0.6,   1. ,   0.2,  -0.6,  -0.8,  -0.2,   0.4,  -2.5,
        -7.3, -12.1,  -8.8, -10.1,  -8.3,  -5.4,  -2.7,   1.5,   4.4,
         0. ,   0.5,   1.5,   1.9,   2.2,   0.4,  -2.5,  -4.6,  -0.7,
        -5.3,  -5.6,  -2. ,  -2.3,   2.1,   1.5,   1.5,   0.9,  -1.8,
        -4.2,  -4.2,  -2.9,   0.2,   1.2,   1.4,   2.2,  -0.9,   0.6,
         0.6,   3.8,   3.4,   2. ,   1.6,   0.6,   0.2,   1.6,   2.3,
         2.5,   0.5,   1.9,   4.8,   6.6,   2.4,   0.4,  -0.4,   0.2,
         1.5,   4.3,   3. ,   3.8,   3.3,   4.7,   4.1,   3.3,   3.7,
         6. ,   4.3,   1.8,   1.2,  -0.5,  -1.4,  -1.7,   0.6,   0.3,
         1.2,   2.5,   5.7,   3.1,   2.9,   3.6,   2.8,   3.2,   4.4,
         4.1,   2.3,   2.2,   7.6,   9.4,   9.2,   6.7,  10.1,  11.1,
         6.1,   4.4,

In [3]:
np.sum(a < 0)

[False  True  True  True  True  True  True  True False False  True  True
 False False  True  True  True False False  True  True False False  True
  True False False False False False False  True  True  True False  True
  True  True  True  True  True  True  True False False False False False
 False False False  True  True  True  True  True  True  True False False
 False False  True  True  True  True False False False False  True False
 False False False False False False False False False False False False
 False False False False  True False False False False False False False
 False False False False False False False  True  True  True False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False False False False False Fa

In [6]:
print("Number of days with the temperature below zero", np.sum(a < 0))


Number of days with the temperature below zero 49


In core Python we can combine truth values using the `and, or, and not` keywords. For boolean array however we have to use the elementwise operators `&, |, and ~`, respectively. An example of these:

In [9]:
print(np.sum((0 < a) & (a > 10))) # Tempeture is greater than 0 and less that 10

128


Note that we had to use parentheses around the comparisons, because the precedence of the `&` is higher than that of `<`, and so it would have been performed before the comparisons.

Another use of boolean arrays is that they can be used to select a subset of elements. For example

In [11]:
c = a > 0
print(c)

[ True False False False False False False False  True  True False False
  True  True False False False  True  True False False  True  True False
 False  True  True  True  True  True  True False False False  True False
 False False False False False False False  True  True False  True  True
  True  True  True False False False False False False False  True  True
  True  True False False False False  True  True  True  True False  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True False  True  True  True  True  True  True  True
  True  True  True  True  True  True  True False False False  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True  T

In [12]:
c.ndim

1

In [14]:
print(c[:10]) # print only the first te elements

[ True False False False False False False False  True  True]


In [19]:
print(a[c]) # select only the positive temperatures

[ 0.6  0.5  1.7  1.1  0.8  1.1  1.6  1.   0.1  1.9  1.6  0.8  0.6  1.
  0.2  0.4  1.5  4.4  0.5  1.5  1.9  2.2  0.4  2.1  1.5  1.5  0.9  0.2
  1.2  1.4  2.2  0.6  0.6  3.8  3.4  2.   1.6  0.6  0.2  1.6  2.3  2.5
  0.5  1.9  4.8  6.6  2.4  0.4  0.2  1.5  4.3  3.   3.8  3.3  4.7  4.1
  3.3  3.7  6.   4.3  1.8  1.2  0.6  0.3  1.2  2.5  5.7  3.1  2.9  3.6
  2.8  3.2  4.4  4.1  2.3  2.2  7.6  9.4  9.2  6.7 10.1 11.1  6.1  4.4
  1.5  1.5  2.9  5.1  7.5  9.8  9.   7.3  9.4 11.2 16.1 16.9 14.7 14.6
 13.3 12.4 14.1 14.1  9.6 13.2 14.1  9.6 10.8  6.5  6.2 10.   9.9 10.4
 14.5 16.4 12.3 14.  16.3 16.8 12.6 12.6 14.8 15.1 18.4 19.3 16.9 17.8
 12.7 12.4 12.2 10.8 12.9 14.  13.5 13.7 15.3 16.3 17.2 14.1 16.4 14.5
 14.1 14.  11.5 14.7 16.1 17.4 18.  16.8 16.8 14.7 15.3 16.7 18.  15.4
 15.  13.7 14.6 15.8 15.6 16.9 15.7 16.1 17.2 19.  19.1 17.8 18.3 17.8
 18.9 17.5 18.2 16.6 17.1 16.  15.2 17.2 18.3 18.7 18.1 19.6 17.4 16.1
 16.4 17.6 19.  18.  18.2 17.4 16.2 14.9 13.5 12.9 11.2 10.4 10.7 12.
 14.4 16

This operation is called `masking`. It can also be used to assign a new value. For example the following zeroes out the negative temperatures:

In [20]:
a[~c] = 0
print(a)

ValueError: assignment destination is read-only

## Boolean Masking — Simply

This is a powerful NumPy technique called **boolean indexing** (or "masking"). It lets you **filter an array** — keeping only the elements that satisfy a condition — without any loop. Let's break down the two operations.

---

### The Setup

```python
a = ... ['Air temperature (degC)'].values     # a 1D array of temperatures
```

So `a` is an array of numbers like `[-2.1, 3.4, 0.0, 5.2, -1.8, ...]` — daily air temperatures, some positive, some negative.

---

### Part 1 — `c = a > 0` Creates a Boolean Array (the "mask")

```python
c = a > 0
```

This is a **vectorized comparison** — it checks `> 0` for **every element at once**, producing an array of `True`/`False` values:

```
a:  [-2.1,  3.4,  0.0,  5.2, -1.8, ...]
         │     │     │     │     │
       -2.1>0 3.4>0 0.0>0 5.2>0 -1.8>0
         │     │     │     │     │
c:  [False, True, False, True, False, ...]
```

**`c` has the same length as `a`**, but every element is a boolean: `True` where the temperature is positive, `False` where it isn't. This boolean array is called a **mask**.

```python
print(c[:10])     # shows the first 10 True/False values
# → [False  True False  True ... ]
```

---

### Part 2 — `a[c]` Uses the Mask to SELECT Elements

```python
print(a[c])
```

This is the clever part: indexing an array **with a boolean array** keeps **only the elements where the mask is `True`**, and drops the rest:

```
a:  [-2.1,  3.4,  0.0,  5.2, -1.8, ...]
c:  [False, True, False, True, False, ...]
        ✗     ✓     ✗     ✓     ✗
             keep        keep

a[c]:  [3.4, 5.2, ...]     ← only the elements where c was True!
```

So `a[c]` returns a **new array containing only the positive temperatures** — every element where the condition `a > 0` held true. Elements where `c` was `False` (negatives and zero) are excluded.

---

### The Two Steps Together

```python
c = a > 0        # STEP 1: build a True/False mask ("which ones are positive?")
a[c]              # STEP 2: use the mask to keep only the True positions
```

You can even combine them into one line:

```python
a[a > 0]          # same result — filter directly, no separate 'c' variable
# → array of only the positive temperatures
```

---

### Why This Is Powerful — No Loop Needed

Without boolean masking, you'd write a loop:

```python
# The slow, manual way:
positives = []
for temp in a:
    if temp > 0:
        positives.append(temp)
```

Boolean masking does the **exact same thing** in one vectorized expression:

```python
positives = a[a > 0]      # one line, no loop, much faster
```

Both give the positive temperatures — but the masking version is shorter and runs in fast compiled code (the vectorization benefit you learned earlier).

---

### Other Things You Can Do With Masks

```python
a[a > 0]          # positive temperatures
a[a < 0]          # negative temperatures (freezing days)
a[a == 0]          # exactly zero
a[(a > 0) & (a < 10)]   # BETWEEN 0 and 10 (& = AND, note the parentheses!)
```

You can also **count** or **compute** on the filtered results:

```python
np.sum(a > 0)      # → how MANY positive days (True counts as 1)
np.mean(a[a > 0])   # → average of just the positive temperatures
len(a[a > 0])        # → number of positive temperatures
```

---

### Step-by-Step on a Tiny Example

```python
a = np.array([-2, 5, 0, 8, -1])

c = a > 0
print(c)          # → [False  True False  True False]

print(a[c])        # → [5 8]     only elements where c is True
```

```
a:      [-2,    5,    0,    8,   -1]
c:      [False, True, False, True, False]
a[c]:         [5,          8      ]    ← kept only the True positions
```

---

### The One-Sentence Summary

> `c = a > 0` creates a **boolean mask** — an array of `True`/`False`, one per element, marking which temperatures are positive; then `a[c]` uses that mask to **select only the elements where `c` is `True`**, returning a new array of just the positive temperatures. This is **boolean indexing** — a fast, loop-free way to filter an array by a condition, equivalent to writing `a[a > 0]` in one step. 🎯

"Indexing an Array" — Simply

Indexing means accessing element(s) from an array by their position or by some selector — reaching into the array to pull out specific values. It's the general term for "getting stuff out of an array using the square brackets [ ]."

The Basic Idea — Square Brackets Reach Into the Array
python
a = np.array([10, 20, 30, 40, 50])

Whenever you use [ ] after an array to grab something from it, that's indexing:

python
a[0]        # → 10    grab the element at position 0
a[2]         # → 30    grab the element at position 2